File: aoc-2019/src/main/java/net/bbelovic/adventofcode/year2019/day1/Day1PuzzlePart2.java
package net.bbelovic.adventofcode.year2019.day1;

import net.bbelovic.adventofcode.Puzzle;

import java.util.List;

public class Day1PuzzlePart2 implements Puzzle<List<String>, Long> {

    private static long getSum(long x) {
        var sum = 0L;
        var z = 0L;
        while ((z = Math.floorDiv(x, 3) - 2) > 0) {
            sum += z;
            x = z;
        }
        return sum;
    }

    @Override
    public Long solve(List<String> input) {
        return input.stream().mapToLong(Long::parseLong)
                .map(Day1PuzzlePart2::getSum)
                .sum();
    }
}

File: aoc-2020/src/main/kotlin/net/bbelovic/adventofcode/year2020/day1/Day1PuzzlePart2.kt
package net.bbelovic.adventofcode.year2020.day1

import net.bbelovic.adventofcode.Puzzle

class Day1PuzzlePart2: Puzzle<List<String>, Long> {
    override fun solve(input: List<String>): Long {
        return Day1PuzzlePart1Recursion(3).solve(input)
    }
}
